element-connector/src/main/java/org/eclipse/californium/elements/PrincipalEndpointContextMatcher.java

```diff
@@ -34,6 +36,15 @@ public String getName() {
 		return "principal correlation";
 	}
 
+	@Override
+        public byte[] getEndpointIdentifier(EndpointContext endpointContext) {
+                Principal principal = endpointContext.getPeerIdentity();
+                if (principal == null) {
+                        throw new IllegalArgumentException("principal must be provided!");
+                }
+                return principal.getName().getBytes();
```

Comment:
I rethink about that this night :p and principal.getName().getBytes() seems not enough to me.
Identifier should contains the type too.  I mean a PreSharedKeyIdentity and RawPublicKeyIdentity are not the same even if both have the same "name".

Reply:
Therefore the matcher will report that violation in `isResponseRelatedToRequest`.

Maybe I move the special `UserInfo` processing than to `isToBeSent`.

Reply:
> Therefore the matcher will report that violation in isResponseRelatedToRequest.

But there is still the issue when we register outbound request.  (in `UdpMatcher.sendRequest`),  or when we register observation (not yet implemented).
I imagine there is poor chance this kind of collision happens but in the meantime this is easy to avoid this just by adding one byte for the type in endpoint identifier.

Reply:
RPK has a special prefix ("ni:///sha-256;"), x509 uses a structured DN.
So I guess, if someone uses a PSK identity, which, matches one of the others, he will also find other ways to croc the system :-).

Reply:
> he will also find other ways to croc the system :-).

I'm not sure, usurp identity seems just super easy in that case...
You are a customer of a iot platform. You sniff traffic and get public key of public certificate of devices. 
Now you register a new device and you choose PSK mode and set the identity with information you get before and you could have same endpoint identifier than devices you sniffed. So all the security is on token generation.

Anyway, if you think this must not be in californium, we are still able to implement this only in Leshan. That's the good point with the way this is implemented :)

Reply:
> Now you register a new device and you choose PSK mode and set the identity with information

FMPOV, it's not possible to support "free PSK identity" selection, because this may cause ambiguous identities. So, if the identities are checked anyway, it's easy to filter out the RPK header and check, that no DN pattern is used.

By the way, AFAIK, leshan could not completely reconstruct the Principal from the registry (without extend it more), therefore I implemented a `Lwm2mEndpointContextMatcher`. From the LWM2M TS, it's clear that only 3 kind of principals are supported and so it's easy to add the prefix at the level of that matcher.

But I found an other small leak, if we use PSK with ServerNameIndication, I guess the Identity name must be extended by that ServerName.